### PR TITLE
demangle: More elaborated demangling for Rust

### PR DESCRIPTION
**Pull request by ChoKyuWon**
This commit fixes following things:
1. Ignore traits (See #1620)
2. Demangle '..' to '::'
3. Consume extra '_' at starting point of string
4. Add placeholder to support Rust v0 mangling rule (start with '_R')

However, because of 2, it replace the '..' to '::', it need a write permission for the source string. It makes segfault for the demangle test because the source string is const.

Signed-off-by: ChoKyuWon <kyuwoncho18@gmail.com>

### CODE REVIEW

File: utils/demangle.c [hunk 1]
```diff
@@ -1415,6 +1415,15 @@ static int dd_source_name(struct demangle_data *dd)
 	dd_append_separator(dd, "::");
 
 	p = dd->old + dd->pos;
+	/* replace ".." to standard separator */
+	for (int i = 0; i < num - 1; ++i) {
+		if (p[i] == '.' && p[i + 1] == '.') {
+			// FIXME: This cause segfault in test because test array is constant.
```
Comment (namhyung) [inline utils/demangle.c:1421]: You cannot change the old string.  Please do it for new string.

**Review by namhyung — COMMENTED** [review]
Thanks for the PR.  Can you please split the ".." part and trait handling?

File: utils/demangle.c [hunk 2]
```diff
@@ -1426,19 +1427,30 @@ static int dd_source_name(struct demangle_data *dd)
 	/* check special symbol mappings (e.g. '$LT$') for Rust */
 	while (dollar != NULL && dollar < end) {
 		bool found = false;
-
 		num = dollar - p;
-		dd_append_len(dd, p, num);
+		/* consume extra '_' before '$' */
```
Comment (namhyung) [inline utils/demangle.c:1432]: I'm not sure about this.  How can we make sure if it's not a valid identifier.  For example it might have `my_mod_` as a module name, then any dollar sign after that would delete the trailing underscore, right?

File: utils/demangle.c [hunk 3]
```diff
@@ -1691,7 +1703,13 @@ static char *demangle_simple(char *str)
 		dd.len -= 15;
 	}
 
-	if (dd.old[0] != '_' || dd.old[1] != 'Z')
+	if (dd.old[0] != '_') {
+		if (dd.old[1] != 'Z' || dd.old[1] != 'R')
+			return xstrdup(str);
+	}
+
+	// TODO: implement demangling Rust v0 mangling
```
Comment (namhyung) [inline utils/demangle.c:1717]: Thanks for doing this.  Could you please create an issue with this?  It'd be great if it can have some examples.

File: utils/demangle.c [hunk 4]
```diff
@@ -1469,6 +1470,11 @@ static int dd_source_name(struct demangle_data *dd)
 	if (add_name)
 		dd_append_len(dd, p, num);
 	__dd_consume_n(dd, num, NULL);
+	separator = strstr(dd->new, "..");
```
Comment (namhyung) [inline utils/demangle.c:1473]: I'm afraid it might have multiple `".."`.  Does this handle it properly?

**Review by namhyung — COMMENTED** [review]
Can you rewrite the header line of the second commit message?  The same message confuses readers and I'd prefer being more specific like 'demangle: convert ".." to "::" for Rust'.

File: utils/demangle.c (hunk 4 above)
Comment (ChoKyuWon) [reply utils/demangle.c:1473]: I'm not sure why, but in my test case, it seems to work properly. But to make sure, I'll add a loop for that.

File: utils/demangle.c (hunk 2 above)
Comment (ChoKyuWon) [reply utils/demangle.c]: That's a good point. I'll revert this code. However, for clearer context, the first underscore would be deleted IMO.

File: utils/demangle.c (hunk 4 above)
Comment (namhyung) [reply utils/demangle.c]: I think it worked by luck.  It just converts a single instance, but later `dd_source_name()` call would convert it again.  If the original name contains more than two ".." and it has a `dd_source_name()` after that, not all ".." would be converted.

File: utils/demangle.c [hunk 5]
```diff
@@ -1464,6 +1465,15 @@ static int dd_source_name(struct demangle_data *dd)
...
 		dd_append_len(dd, p, num);
 	__dd_consume_n(dd, num, NULL);
+	while (true) {
+		separator = strstr(dd->new, "..");
+		if (separator) {
+			separator[0] = ':';
+			separator[1] = ':';
+		}
+		else
+			break;
```
Comment (namhyung) [inline utils/demangle.c:1475]: In general, I prefer reducing the nesting level by checking the break condition early.  Something liike below..

```c
while (true) {
    separator = strstr(dd->new, "..");
    if (separator == NULL)
        break;
    separator[0] = ':';
    separator[1] = ':';
}
```

File: utils/demangle.c [hunk 6]
```diff
@@ -1382,6 +1382,7 @@ static int dd_source_name(struct demangle_data *dd)
 {
 	int num = dd_number(dd);
 	char *dollar;
+	char *separator;
```
Comment (namhyung) [inline utils/demangle.c:1385]: You may put it inside the loop below.

File: utils/demangle.c (hunk 4 above)
Comment (ChoKyuWon) [reply utils/demangle.c]: You're right. Thanks for the review.

File: utils/demangle.c [hunk 7]
```diff
@@ -1464,6 +1465,15 @@ static int dd_source_name(struct demangle_data *dd)
...
 	__dd_consume_n(dd, num, NULL);
+	while (true) {
+		separator = strstr(dd->new, "..");
+		if (separator) {
+			separator[0] = ':';
+			separator[1] = ':';
+		}
+		else
+			break;
+	}
```
Comment (namhyung) [inline utils/demangle.c:1476]: Another concern is that it's suboptimal.  It checks ".." whenever it calls `dd_source_name()`.  I think we can do it only when the dollar sign was found.  Also it doesn't need to search the whole (new) string.  Could we check the newly added region only?

**Review by namhyung — COMMENTED** [review]
Please add some explanation in the second commit message.  Like when it needs to be converted or just include an example.

Also I think we should add a test for more complex case like in the #1620.

File: utils/demangle.c (hunk 7 above)
Comment (ChoKyuWon) [reply utils/demangle.c:1474]: I'm not sure that the ".." is always inside the <>. I'll try to find a counterexample, and if there's nothing, I'll fix this commit.

Comment (ChoKyuWon) [reply utils/demangle.c:1474]: I collect the LLVM IR of large-scale projects and search the pattern like this:
```
$ cat *.ll | grep "define[^w]*\.\." | grep -v "$"
```
and it returns nothing.
So ".." only generated with a dollar sign is quite a safe guess. I'll work on the optimal solution.

File: utils/demangle.c [hunk 8]
```diff
@@ -1426,7 +1426,6 @@ static int dd_source_name(struct demangle_data *dd)
 	/* check special symbol mappings (e.g. '$LT$') for Rust */
 	while (dollar != NULL && dollar < end) {
 		bool found = false;
-
```
Comment (namhyung) [inline utils/demangle.c:1429]: Unnecessary change.  Please leave it.

File: utils/demangle.c [hunk 9]
```diff
@@ -1930,8 +1943,11 @@ TEST_CASE(demangle_rust1)
 	DEMANGLE_TEST("_ZN35Bar$LT$$u5b$u32$u3b$$u20$4$u5d$$GT$E", "Bar<[u32; 4]>");
 	DEMANGLE_TEST("_ZN71_$LT$Test$u20$$u2b$$u20$$u27$static"
 		      "$u20$as$u20$foo..Bar$LT$Test$GT$$GT$3barE",
-		      "_<Test + 'static as foo..Bar<Test>>::bar");
+		      "_<Test + 'static>::bar");
 	DEMANGLE_TEST("_ZN3foo3bar17h05af221e174051e9E", "foo::bar");
+	DEMANGLE_TEST(
+		"_ZN61_$LT$$RF$std..io..stdio..Stdout$u20$as$u20$std..io..Write$GT$9write_fmt17h75c561f414a62159E",
+		"_<&std::io::stdio::Stdout>::write_fmt");
```
Comment (namhyung) [inline utils/demangle.c:1959]: This should be moved to the next commit.

File: utils/demangle.c [hunk 10]
```diff
@@ -1426,8 +1426,17 @@ static int dd_source_name(struct demangle_data *dd)
 	/* check special symbol mappings (e.g. '$LT$') for Rust */
 	while (dollar != NULL && dollar < end) {
 		bool found = false;
+		char *separator = p;
```
Comment (namhyung) [inline utils/demangle.c:1429]: The general guide is to have a blank line between declaration and body.

**Review by namhyung — COMMENTED** [review]
Thanks for your work.  I only have a few nitpicks.

**Review by namhyung — APPROVED** [review]
LGTM. Can you please rebase it onto the current master?

**Comment by namhyung** [comment]
You need to add a blank line after the commit header to separate subject and body.  Please take a look at other commit messages.
